aten/src/ATen/core/TensorOptions.h

```diff
@@ -132,54 +130,54 @@ struct CAFFE2_API TensorOptions {
   }
 
   /// Returns the device of the `TensorOptions`.
-  const Device& device() const noexcept {
-    return device_;
-  }
+  Device device() const noexcept;
+
+  optional<Device> device_opt() const noexcept { return device_; }
 
   /// Returns the device index of the `TensorOptions`.
   int32_t device_index() const noexcept {
-    return device_.index();
+    return device().index();
   }
 
   /// Returns the dtype of the `TensorOptions`.
-  ScalarType dtype() const noexcept {
-    return dtype_;
-  }
+  ScalarType dtype() const noexcept;
+
+  optional<ScalarType> dtype_opt() const noexcept { return dtype_; }
 
   /// Returns the layout of the `TensorOptions`.
-  Layout layout() const noexcept {
-    return layout_;
-  }
+  Layout layout() const noexcept;
+
+  optional<Layout> layout_opt() const noexcept { return layout_; }
 
   /// Returns the `requires_grad` property of the `TensorOptions`.
-  bool requires_grad() const noexcept {
-    return requires_grad_;
-  }
+  bool requires_grad() const noexcept;
+
+  optional<bool> requires_grad_opt() const noexcept { return requires_grad_; }
 
   /// Returns the `is_variable` property of the `TensorOptions`.
-  bool is_variable() const noexcept {
-    return is_variable_;
-  }
+  bool is_variable() const noexcept;
+
+  optional<bool> is_variable_opt() const noexcept { return is_variable_; }
 
   // Resolves the ATen backend specified by the current construction axes.
   Backend backend() const noexcept {
```

Comment:
Do we still need this method? 🤔

Reply:
I'm not sure how we would not need it (in principle, anyway) :) This must stay until we merge Variable and Tensor.